v2/internal/util/postgresql/psqluser.go

```diff
@@ -0,0 +1,167 @@
+func ConnectToDB(ctx context.Context, fullservername string, database string, port int, user string, password string) (*sql.DB, error) {
+
+	connString := fmt.Sprintf("host=%s user=%s password=%s port=%d dbname=%s sslmode=require connect_timeout=30", fullservername, user, password, port, database)
+
+	db, err := sql.Open(PDriverName, connString)
+	if err != nil {
+		return db, err
+	}
+
+	err = db.PingContext(ctx)
+	if err != nil {
+		return db, err
+	}
+
+	return db, err
+}
+
+// ReconcileGrantUserRoles grants roles to a user for a given database
+func ReconcileGrantUserRoles(ctx context.Context, db *sql.DB, user string, roles []string) error {
+	var errorStrings []string
+
+	if err := FindBadChars(user); err != nil {
+		return errors.Wrap(err, "problem found with username")
+	}
+
+	for _, role := range roles {
+		tsql := fmt.Sprintf("GRANT %s TO %q", role, user)
+
+		if err := FindBadChars(role); err != nil {
+			return errors.Wrap(err, "problem found with role")
+		}
+
+		_, err := db.ExecContext(ctx, tsql)
+		if err != nil {
+			errorStrings = append(errorStrings, err.Error())
+		}
+	}
+
+	if len(errorStrings) != 0 {
+		return fmt.Errorf(strings.Join(errorStrings, "\n"))
+	}
+	return nil
+}
+
+func CreateOrUpdateUser(ctx context.Context, db *sql.DB, username string, password string) error {
+	// make an effort to prevent sql injection
+	if err := FindBadChars(username); err != nil {
+		return errors.Wrap(err, "problem found with username")
+	}
+	if err := FindBadChars(password); err != nil {
+		return errors.Wrap(err, "problem found with password")
+	}
+
+	// we call both CREATE and ALTER here so achieve an idempotent operation that also updates the password seamlessly
+	// if it has changed
+
+	tsql := fmt.Sprintf("CREATE USER \"%s\" WITH PASSWORD '%s'", username, password)
+	_, err := db.ExecContext(ctx, tsql)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create user %s", username)
+	}
+
+	tsql = fmt.Sprintf("ALTER USER '%s' WITH PASSWORD '%s'", username, password)
+	_, err = db.ExecContext(ctx, tsql)
+	if err != nil {
+		return errors.Wrapf(err, "failed to alter user %s", username)
+	}
+
+	return nil
+}
+
+// DoesUserExist checks if db contains user
+func DoesUserExist(ctx context.Context, db *sql.DB, username string) (bool, error) {
+	res, err := db.ExecContext(ctx, "SELECT * FROM pg_user WHERE usename = $1", username)
+	if err != nil {
+		return false, err
+	}
+	rows, err := res.RowsAffected()
+	return rows > 0, err
+
+}
+
+// DropUser drops a user from db
+func DropUser(ctx context.Context, db *sql.DB, user string) error {
+	if err := FindBadChars(user); err != nil {
+		return errors.Wrap(err, "problem found with username")
+	}
+
+	tsql := fmt.Sprintf("DROP USER IF EXISTS %q", user)
+	_, err := db.ExecContext(ctx, tsql)
+	return err
+}
+
+// GetUserRoles gets the user roles the user has as a set.
+func GetUserRoles(ctx context.Context, db *sql.DB, user string) (set.Set[string], error) {
```

Comment:
I don't see where this is called except in the tests? I _think_ we need this combined with something like we have in MySQL for doing role diffing. See https://github.com/Azure/azure-service-operator/blob/main/v2/internal/util/mysql/privilege.go#L29